File: db4oj/core/src/com/db4o/inside/slots/ReferencedSlot.java
/* Copyright (C) 2004   db4objects Inc.   http://www.db4o.com */

package com.db4o.inside.slots;

import com.db4o.*;

public final class ReferencedSlot extends Slot{
    
	public int _references;
	
	public ReferencedSlot(int address, int length){
        super(address, length);
	}
	
	public Object read(YapReader a_bytes){
		int address = a_bytes.readInt();
		int length = a_bytes.readInt();
		return new ReferencedSlot(address, length);
	}
}

File: db4on/core/src/core/com/db4o/inside/slots/ReferencedSlot.cs
namespace com.db4o.inside.slots
{
	/// <exclude></exclude>
	public class ReferencedSlot : com.db4o.TreeInt
	{
		private com.db4o.inside.slots.Slot _slot;

		private int _references;

		public ReferencedSlot(int a_key) : base(a_key)
		{
		}

		public override object ShallowClone()
		{
			com.db4o.inside.slots.ReferencedSlot rs = new com.db4o.inside.slots.ReferencedSlot
				(_key);
			rs._slot = _slot;
			rs._references = _references;
			return base.ShallowCloneInternal(rs);
		}

		public virtual void PointTo(com.db4o.inside.slots.Slot slot)
		{
			_slot = slot;
		}

		public virtual com.db4o.foundation.Tree Free(com.db4o.YapFile file, com.db4o.foundation.Tree
			 treeRoot, com.db4o.inside.slots.Slot slot)
		{
			file.Free(_slot._address, _slot._length);
			if (RemoveReferenceIsLast())
			{
				return treeRoot.RemoveNode(this);
			}
			PointTo(slot);
			return treeRoot;
		}

		public virtual bool AddReferenceIsFirst()
		{
			_references++;
			return (_references == 1);
		}

		public virtual bool RemoveReferenceIsLast()
		{
			_references--;
			return _references < 1;
		}

		public virtual com.db4o.inside.slots.Slot Slot()
		{
			return _slot;
		}
	}
}
